### PR TITLE
Calculate status in a separate thread

### LINKED ISSUE #5755: get_status for surface (and probably other layers) should be monitored, debounced/throttled, and run on a separate thread
## 🧰 Task

In [this discussion](https://github.com/napari/napari/pull/5642#issuecomment-1515964705), @brisvag and @aganders3 figured out that the [texture example](https://github.com/napari/napari/blob/main/examples/surface_multi_texture.py) had low framerate not because of VisPy and OpenGL, but because of layer.get_status. The simple workaround for now is to not have the layer selected when interacting with it, but ideally we should:

1. Make sure that get_status runs in a separate thread so it doesn't interfere with the GUI responsiveness
2. Monitor how long a get_status call takes, and
3. Use debounce and/or throttle (probably debounce) to make sure it doesn't get called more than that.

I prefer debounce, at least for long-running get_status, because if get_status takes a while to run, then throttling it will only show an outdated status message anyway, defeating the purpose of it and wasting precious CPU cycles.

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

Closes #5755 

# Description

Currently, status bar updates are calculated in the main thread any time the cursor moves over the canvas, which can cause a laggy UI when the calculation is slow. This PR moves the calculation to a separate thread for all layers.

To avoid overwhelming the status thread, a bool flag is used to deduplicate requests.

### CODE REVIEW

File: napari/_qt/threads/status_checker.py
Comment (jni) [inline napari/_qt/threads/status_checker.py]: I don't like not re-raising here...

**Comment by jni** [comment]
Cool, very exciting @Czaki!

I'm not 100% on this:

> This PR removes update status from ViewerModel and move signal connection to _QtMainWindow, as its calculation is useless until the status bar exists.

Isn't it true that other users could connect to this event and grab the status, even if we don't have a mainwindow?

File: napari/_qt/threads/status_checker.py
Comment (Czaki) [reply napari/_qt/threads/status_checker.py]: Reraising will stop thread. There is propagation of exception to notification manager. I may also try call `sys.excepthook` or print exception

**Comment by Czaki** [comment]
> Isn't it true that other users could connect to this event and grab the status, even if we don't have a mainwindow?

Yes. But also is someone create only `QtViewer` widget and never touch status text. And still be negatively affected by performance. We may want to create a public API and document how to enable this if a user wants.

File: napari/_qt/threads/status_checker.py
Comment (jni) [reply napari/_qt/threads/status_checker.py]: But I *want* it to stop the thread because otherwise we are masking potentially fatal bugs? And it becomes impossible to drop into the traceback with a debugger?

**Comment by jni** [comment]
> We may want to create a public API and document how to enable this if a user wants.

As long as we can avoid an overly complex solution... Something that can be controlled by an API and maybe a preference...

**Comment by jni** [comment]
@Czaki I love this feature but since we want to release 0.5.2 this week and this looks like it'll require a bit of discussion, I'm going to bump it until 0.5.3.

**Comment by Czaki** [comment]
> this looks like it'll require a bit of discussion

This requires to debug why min_req tests are hanging

File: napari/_qt/threads/status_checker.py
Comment (Czaki) [reply napari/_qt/threads/status_checker.py]: It will be same as current fail to get status. We currently also do not break the main thread but just pass exception to notification manager.

**Comment by Czaki** [comment]
@jni I Think that merge of #7158 will fix CI problem in this PR. This will allow merging this PR into 0.5.2 And allow removing a small workaround from #7144 that globally will bring big improvement in usage of napari with shapes and surfaces.

**Comment by brisvag** [comment]
> This PR removes update status from ViewerModel and move signal connection to `_QtMainWindow`, as its calculation is useless until the status bar exists.

`status` and `tooltip` are public API on the `ViewerModel`, which I don't think we should remove here (if at all). I think we should still make sure to update the model rather than just the qt window.

Also, should we combine throttling with threading?

**Comment by Czaki** [comment]
> `status` and `tooltip` are public API on the `ViewerModel`, which I don't think we should remove here (if at all). I think we should still make sure to update the model rather than just the qt window.

I do not remove it. I just removed updating its status based on cursor position until spawn main window.

**Comment by Czaki** [comment]
> Yeah but we should keep them up to date still, I think.

Why?

**Comment by brisvag** [comment]
This might be updated via the model in headless mode by changing the cursor. Not sure if there's any use cases in the wild, but I don't see why we should suddenly relegate this just to the Qt viewer when so much else isn't qt-only.

**Comment by Czaki** [comment]
> This might be updated via the model in headless mode by changing the cursor. Not sure if there's any use cases in the wild, but I don't see why we should suddenly relegate this just to the Qt viewer when so much else isn't qt-only.

Then they still call `update_status_from_cursor` method and get updated state of viewer model based on cursor position. 

The performance impact of this operation is colossal, and keeping it because of some potential edge case is not worth it.

**Comment by brisvag** [comment]
> The performance impact of this operation is colossal, and keeping it because of some potential edge case is not worth it.

Ok, color me convinced :+1:

File: pyproject.toml
Comment (jni) [inline pyproject.toml]: Why these changes?

File: tox.ini
Comment (Czaki) [inline tox.ini]: @jni I have moved to triggering parallel mode here. It simplifies local coverage calculation. I may move this to a separate PR.

Comment (jni) [reply tox.ini]: Yeah I would prefer that. It makes it easier to think about this PR — I was wondering if there was some change in the parallelism constraints given that there is some new threading code.

**Comment by jni** [comment]
> The performance impact of this operation is colossal, and keeping it because of some potential edge case is not worth it.

To be honest, I would prefer to have things happen in two PRs:
- calculate status in separate thread
- only update status automatically from Qt

This would isolate the potentially breaking change for some use cases to a single PR.

Having said this, I recognise it's some extra work, for a potentially small gain, so I am happy to let this through if you are too time-constrained @Czaki.

**Review by jni — APPROVED** [review]

**Review by brisvag — APPROVED** [review]

**Comment by Czaki** [comment]
@jni 
I have reverted the change of disabling update of status in viewer model. I cannot open PR with this change until this PR is merged

File: napari/_qt/qt_main_window.py
Comment (jni) [inline napari/_qt/qt_main_window.py]:
```suggestion
    def _toggle_status_thread(self, event: Event):
        if event.value:
```

File: napari/conftest.py
Comment (jni) [inline napari/conftest.py]: Ok I was trying to understand these... Here's my understanding:

- the names "my_run" and "my_constructor" are arbitrary — we could use more descriptive names, right?
- if so, I would propose `_init_with_trace` and `_run_with_trace`
- is assigning `base_constructor` needed? Could we not replace that line (443) with `QThread.__init__(self, *args, **kwargs)`?
- and add the following docstrings:
  - for `_init_with_trace`:
    ```
    """Constructor for QThread adding tracing for coverage measurements.

    Functions running in QThreads don't get measured by coverage.py, see
    https://github.com/nedbat/coveragepy/issues/686. Therefore, we will
    monkeypatch the constructor to add to the thread to `sys.settrace` when
    we call `run` and `coverage` is in `sys.modules`.
    """
    ```
  - for `_run_with_trace`:
    ```
    """QThread.run but adding execution to sys.settrace when measuring coverage.

    See https://github.com/nedbat/coveragepy/issues/686#issuecomment-634932753
    and `_init_with_trace`. When running QThreads during testing, we monkeypatch
    the QThread constructor and run methods with traceable equivalents.
    """
    ```

(I'd also rename `_base_run` to `_qthread_run`)

Comment (jni) [reply napari/conftest.py]: (Is my understanding correct and if so what do you think about these suggestions @Czaki?)

File: napari/_qt/threads/status_checker.py
Comment (jni) [inline napari/_qt/threads/status_checker.py]: @Czaki as far as I can tell, `_need_status_update` and `_event` are redundant? (Whenever `_event` is set, `_need_status_update` is set to `True`, and whenever `_event` is cleared, `_need_status_update` is set to `False`?) So, if my understanding is right, I would change the code as follows:

```suggestion
class StatusChecker(QThread):
    """Separate thread dedicated to updating the status bar.
    
    Prior to https://github.com/napari/napari/pull/7146, status bar updates
    happened on the main thread in the viewer model, which could be
    expensive in some layers and resulted in bad performance when some
    layers were selected.
    
    This class puts the job of computing the status into a separate thread,
    which then uses Qt Signals to update the main viewer with the status
    string.
    
    Because the thread runs a single infinite while loop, the updates are
    naturally throttled to as fast as they can be computed, but no faster.
    
    Attributes
    ----------
    _need_status_update : threading.Event
        An Event (fancy thread-safe bool) for keeping track of when the
        status needs updating (because the cursor has moved).
    _terminate : bool
        Whether the thread needs to be terminated. Set by the QtViewer
        when it is being closed. No more status updates will take place if
        _terminate is set.
    viewer_ref : weakref.ref[napari.viewer.ViewerModel]
        A weak reference to the viewer providing the status updates. We
        don't want this thread to prevent the viewer from being garbage
        collected, so we keep only a weak reference, and check it when
        we need to compute a new status update.
    """

    status_and_tooltip_changed = Signal(object)

    def __init__(self, viewer: ViewerModel, parent: QObject | None = None):
        super().__init__(parent=parent)
        self.viewer_ref = ref(viewer)
        self._need_status_update = Event()
        self._need_status_update.clear()
        self._terminate = False

    def trigger_status_update(self) -> None:
        self._need_status_update.set()

    def terminate(self) -> None:
        self._terminate = True
        self._need_status_update.set()

    def run(self) -> None:
        while not self._terminate:
            if self._need_status_update.is_set():
                self._need_status_update.clear()
                self.calculate_status()
            else:
                self._need_status_update.wait()
```

(Whether or not my understanding is correct, some version of the docstring should be added...)

File: napari/_qt/threads/status_checker.py
Comment (jni) [reply napari/_qt/threads/status_checker.py]: I don't see the exception passing being removed (I would expect a removed line containing `notification_manager`?), but I don't think I want to hold up the PR for this discussion, happy to postpone it.

File: napari/conftest.py
Comment (Czaki) [reply napari/conftest.py]: > is assigning `base_constructor` needed? Could we not replace that line (443) with `QThread.__init__(self, *args, **kwargs)`?

It may not work if more than one fixtures will try to do something with QThread constructor.

File: napari/_qt/threads/status_checker.py
Comment (Czaki) [reply napari/_qt/threads/status_checker.py]: Looks like it is obsolete. I try to create deadlock but fail

File: napari/_qt/threads/status_checker.py
Comment (Czaki) [reply napari/_qt/threads/status_checker.py]: @jni I have added a comment in `except` clause to clarify why I want to only propagate exception to the notification manager.

**Comment by psobolewskiPhD** [comment]
Hmm, i wanted to try this with the `surface_multi_texture.py` example.
It runs, but choppy on main.
With this branch I get:
`Job 1, 'python examples/surface_multi_t…' terminated by signal SIGBUS (Misaligned address error)`
```
napari: 0.5.0a2.dev359+g0907fdad5
Platform: macOS-14.5-arm64-arm-64bit
System: MacOS 14.5
Python: 3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:13:44) [Clang 16.0.6 ]
Qt: 6.7.1
PyQt6: 
NumPy: 1.26.1
SciPy: 1.14.0
Dask: 2024.7.1
VisPy: 0.14.3
magicgui: 0.8.3
superqt: 0.6.7
in-n-out: 0.2.1
app-model: 0.2.8
npe2: 0.7.7

OpenGL:
  - GL version:  2.1 Metal - 88.1
  - MAX_TEXTURE_SIZE: 16384
  - GL_MAX_3D_TEXTURE_SIZE: 2048

Screens:
  - screen 1: resolution 1680x1050, scale 2.0

Optional:
  - numba: 0.60.0
  - triangle not installed
  - napari-plugin-manager: 0.1.0a3.dev189+gc48deb2

Settings path:
  - /Users/piotrsobolewski/Library/Application Support/napari/napari-dev_a1eb8b76ba95fa16ad06e26097b46b8455dfbf0b/settings.yaml
Plugins:
  - napari: 0.5.1.dev6+g62ef757a5 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.2.0 (2 contributions)
  
```

**Review by psobolewskiPhD — DISMISSED** [review]
crash on surface_multi_texture example:
`Job 1, 'python examples/surface_multi_t…' terminated by signal SIGBUS (Misaligned address error)`
with `pip install numpy==1.26.1`

Looks like we dont have numpy < 2 on CI?

I'm flagging this, but if it's a false alarm, feel free to dismiss. There are plenty of plugins that are not numpy2.

**Comment by jni** [comment]
Wow. 😢

**Comment by jni** [comment]
(Thanks @psobolewskiPhD for testing!)

**Comment by Czaki** [comment]
> Looks like we dont have numpy < 2 on CI?
> 
> I'm flagging this, but if it's a false alarm, feel free to dismiss. There are plenty of plugins that are not numpy2.

Ehhhh. Again... 
I do not see where here is array allocation but looks like NumPy<2 on macos arm hides more problems...
(I also do not check surface code, maybe there is some array inversion). 

Could you test with `numpy<2` but installed from conda?

**Comment by jni** [comment]
@Czaki this issue seems like a difficult fix... Should we postpone to 0.5.3? It's a great contribution but it's only a two week delay

**Comment by Czaki** [comment]
> @Czaki this issue seems like a difficult fix... Should we postpone to 0.5.3? It's a great contribution but it's only a two week delay

I have already fixed this in PartSeg few months ago. I had hope that someone else confirm that this is only problem with numpy from wheels, to just copy the code for checking if it is a problematic numpy version.

**Comment by Czaki** [comment]
Relevant PartSeg pull request https://github.com/4DNucleome/PartSeg/pull/1124

**Comment by Czaki** [comment]
relevant numpy issue https://github.com/numpy/numpy/issues/21799

**Comment by jni** [comment]
> But if you will not be happy with my solution, you could postpone it.

Ah, I'm glad you already had a solution — that wasn't clear from your initial message.

I don't mind it, but I don't love it — as far as I can tell, the whole new thread becomes redundant, so I would expect that we can avoid altogether the creation of the StatusThread and just compute the status in the main thread as we do now? (If I remember correctly, ensure_main_thread is not free, so this PR would now be a performance regression for numpy<2?)

**Comment by psobolewskiPhD** [comment]
I checked 1.26.4 wheel from pypi, same error.
Using `--no-binary numpy` solves the issue as it now builds with Apple Accelerate instead of openblas:
```
In [3]: np.show_config(mode='dicts')['Build Dependencies']['blas']
Out[3]: 
{'name': 'accelerate',
 'found': True,
 'version': 'unknown',
 'detection method': 'system',
 'include directory': 'unknown',
 'lib directory': 'unknown',
 'openblas configuration': 'unknown',
 'pc file directory': 'unknown'}
```
(see also: https://github.com/napari/napari/issues/5040#issuecomment-1874117253 )

Accelerate is also used in 2.0 wheels. 
Does anyone know if numpy may do a 1.26.5 release with Accelerate wheels?

This reminded me of an older numpy/scipy/blas issue, which could be resolved using `OPENBLAS_NUM_THREADS=1`
and indeed that works with pypi wheel.

**Comment by jni** [comment]
> This reminded me of an older numpy/scipy/blas issue, which could be resolved using OPENBLAS_NUM_THREADS=1
and indeed that works with pypi wheel.

That's actually huge!

**Comment by jni** [comment]
@psobolewskiPhD to clarify — are you seeing the crash even with @Czaki's latest commits?

**Comment by psobolewskiPhD** [comment]
With current state of this PR and numpy 1.26.1 (so no accelerate) I get:
```
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
File ~/Dev/napari/napari/_qt/threads/status_checker.py:98, in StatusChecker.calculate_status(self=<napari._qt.threads.status_checker.StatusChecker object>)
     94 try:
     95     if PROBLEMATIC_NUMPY_MACOS:
     96         # On macOS with numpy<2, we need to calculate the status in
     97         # the main thread to avoid a bus error.
---> 98         res = calc_status_in_main_thread(viewer)
        viewer = Viewer(camera=Camera(center=(1.4905714988708496, 0.3298051357269287, 4.073935776948929), zoom=41.0801987132787, angles=(0.0, 0.0, 89.99999999999999), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(1.0, 1.0, 0.0), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=1.0), dims=Dims(ndim=3, ndisplay=3, order=(0, 1, 2), axis_labels=('0', '1', '2'), rollable=(True, True, True), range=(RangeTuple(start=-2.001523971557617, stop=4.982666969299316, step=1.0), RangeTuple(start=-6.60784387588501, stop=7.267454147338867, step=1.0), RangeTuple(start=-0.4753879904747009, stop=8.623259544372559, step=1.0)), margin_left=(0.0, 0.0, 0.0), margin_right=(0.0, 0.0, 0.0), point=(0.9984760284423828, -0.6078438758850098, 3.524612009525299), last_used=0), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Surface layer 'Texture_0' at 0x1377e73e0>, <Surface layer 'GeneratedMat2' at 0x1377e6d80>], help='use <2> for transform', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=False, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x302d95bc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, _keymap={})
     99     else:
    100         res = viewer._calc_status_from_cursor()

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.12/site-packages/superqt/utils/_ensure_thread.py:97, in ensure_main_thread.<locals>._out_func.<locals>._func(_max_args_=1, *args=(Viewer(camera=Camera(center=(1.4905714988708496,...use_drag_gen={}, _mouse_wheel_gen={}, _keymap={}),), **kwargs={})
     95 @wraps(func_)
     96 def _func(*args, _max_args_=max_args, **kwargs):
---> 97     return _run_in_thread(
        func_ = <function calc_status_in_main_thread at 0x3075882c0>
        QCoreApplication = <class 'PyQt6.QtCore.QCoreApplication'>
        await_return = True
        timeout = 1000
        args[:_max_args_] = (Viewer(camera=Camera(center=(1.4905714988708496, 0.3298051357269287, 4.073935776948929), zoom=41.0801987132787, angles=(0.0, 0.0, 89.99999999999999), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(1.0, 1.0, 0.0), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=1.0), dims=Dims(ndim=3, ndisplay=3, order=(0, 1, 2), axis_labels=('0', '1', '2'), rollable=(True, True, True), range=(RangeTuple(start=-2.001523971557617, stop=4.982666969299316, step=1.0), RangeTuple(start=-6.60784387588501, stop=7.267454147338867, step=1.0), RangeTuple(start=-0.4753879904747009, stop=8.623259544372559, step=1.0)), margin_left=(0.0, 0.0, 0.0), margin_right=(0.0, 0.0, 0.0), point=(0.9984760284423828, -0.6078438758850098, 3.524612009525299), last_used=0), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Surface layer 'Texture_0' at 0x1377e73e0>, <Surface layer 'GeneratedMat2' at 0x1377e6d80>], help='use <2> for transform', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=False, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x302d95bc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, _keymap={}),)
        args = (Viewer(camera=Camera(center=(1.4905714988708496, 0.3298051357269287, 4.073935776948929), zoom=41.0801987132787, angles=(0.0, 0.0, 89.99999999999999), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(1.0, 1.0, 0.0), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=1.0), dims=Dims(ndim=3, ndisplay=3, order=(0, 1, 2), axis_labels=('0', '1', '2'), rollable=(True, True, True), range=(RangeTuple(start=-2.001523971557617, stop=4.982666969299316, step=1.0), RangeTuple(start=-6.60784387588501, stop=7.267454147338867, step=1.0), RangeTuple(start=-0.4753879904747009, stop=8.623259544372559, step=1.0)), margin_left=(0.0, 0.0, 0.0), margin_right=(0.0, 0.0, 0.0), point=(0.9984760284423828, -0.6078438758850098, 3.524612009525299), last_used=0), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Surface layer 'Texture_0' at 0x1377e73e0>, <Surface layer 'GeneratedMat2' at 0x1377e6d80>], help='use <2> for transform', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=False, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x302d95bc0>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, _keymap={}),)
        _max_args_ = 1
        kwargs = {}
     98         func_,
     99         QCoreApplication.instance().thread(),
    100         await_return,
    101         timeout,
    102         args[:_max_args_],
    103         kwargs,
    104     )

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.12/site-packages/superqt/utils/_ensure_thread.py:189, in _run_in_thread(func=<function calc_status_in_main_thread>, thread=<PyQt6.QtCore.QThread object>, await_return=True, timeout=1000, args=(Viewer(camera=Camera(center=(1.4905714988708496,...use_drag_gen={}, _mouse_wheel_gen={}, _keymap={}),), kwargs={})
    187 f.finished.connect(future.set_result, Qt.ConnectionType.DirectConnection)
    188 QMetaObject.invokeMethod(f, "call", Qt.ConnectionType.QueuedConnection)  # type: ignore
--> 189 return future.result(timeout=timeout / 1000) if await_return else future
        future = <Future at 0x3072808c0 state=pending>
        await_return = True
        timeout / 1000 = 1.0
        timeout = 1000

File ~/Dev/miniforge3/envs/napari-dev/lib/python3.12/concurrent/futures/_base.py:458, in Future.result(self=None, timeout=1.0)
    456             return self.__get_result()
    457         else:
--> 458             raise TimeoutError()
    459 finally:
    460     # Break a reference cycle with the exception in self._exception
    461     self = None

TimeoutError:
```

Maybe instead of running on main for macOS 1.26 we could check for:
`np.show_config(mode='dicts')['Build Dependencies']['blas']['name'] == 'openblas64'
and if so do `os.environ['OPENBLAS_NUM_THREADS']='1'`? 
Not sure how to reload numpy though...

**Comment by psobolewskiPhD** [comment]
OK, this works in napari __init__:
```
from napari.utils._check_numpy_version import PROBLEMATIC_NUMPY_MACOS
if PROBLEMATIC_NUMPY_MACOS:
    import threadpoolctl
    print("Setting threadpool limits")
    threadpoolctl.threadpool_limits({'libopenblas': 1})
```

**Comment by Czaki** [comment]
> I tried without threadpoolctl but couldn't get it to work, it's trickier because scipy also uses openblas, so finding the library is a pain.

I know how to find a library. Just need to know how to limit threads for BLAS. But checking the code of threapoolctl should help.

**Comment by Czaki** [comment]
@jni @psobolewskiPhD I have implemented solution of setting the number of threads for NumPy openblas. I do not touch SciPy one, as I do not see a report connected with it (and it may be a different version).

I have verified that it works on my macOS. 

It may introduce some negative performance impact on users of numpy<2, as all BLAS operations in the main thread will be executed using a single thread. 
But they will also benefit from the performance of better querying of status.

**Comment by psobolewskiPhD** [comment]
With the latest commits everything works on all 3 versions of numpy I tested (1.26.1, 1.26.6, and 2.0.1), as fresh wheels from pypi. I dont see a difference in performance between the 3. But a nice boost vs main. Also on main when tossing the model around I see ~600% CPU with this it's ~100%.

File: napari/_check_numpy_version.py
Comment (psobolewskiPhD) [inline napari/_check_numpy_version.py]: As the issue is specific to OpenBLAS is it worth also checking for that?
something like `np.show_config(mode='dicts')['Build Dependencies']['blas']['name'] == 'openblas64'`
Then again is a `cibw-run` wheel but without openblas likely/possible?

Relevant: https://github.com/OpenMathLib/OpenBLAS/wiki/faq#how-can-i-use-openblas-in-multi-threaded-applications

Actually maybe that link belongs in the other function.

Comment (psobolewskiPhD) [reply napari/_check_numpy_version.py]: And are we sure this threading thing is arm64 only? Or does x86 numpy use the intel blas...

File: napari/_check_numpy_version.py
Comment (psobolewskiPhD) [inline napari/_check_numpy_version.py:56]:
```suggestion
    This function is loading openblas library from numpy and set number of threads to 1.
    See: https://github.com/OpenMathLib/OpenBLAS/wiki/faq#how-can-i-use-openblas-in-multi-threaded-applications
```

**Comment by jni** [comment]
Question — does the order of imports matter? ie will this approach fail if someone has already imported NumPy?

File: napari/_check_numpy_version.py
Comment (Czaki) [reply napari/_check_numpy_version.py]: > something like `np.show_config(mode='dicts')['Build Dependencies']['blas']['name'] == 'openblas64'`
> Then again is a `cibw-run` wheel but without openblas likely/possible?

This is an issue with a specific version of openblas used by numpy when compiling wheels. So decided to check build information. As someone may have own compilation against more modern openblas

> And are we sure this threading thing is arm64 only? Or does x86 numpy use the intel blas...

On CI, only the arm64 version crashes.

**Comment by Czaki** [comment]
> Question — does the order of imports matter? ie will this approach fail if someone has already imported NumPy?

Order of imports matter. If NumPy is not imported, then `blas = ctypes.CDLL(str(blas_lib[0]), mode=os.RTLD_NOLOAD)` will not load shared library, and we do not get access to proper function. Thisi is one of the reasons why NumPy is imported on being of this file.

**Comment by Czaki** [comment]
> With the latest commits everything works on all 3 versions of numpy I tested (1.26.1, 1.26.6, and 2.0.1), as fresh wheels from pypi. I dont see a difference in performance between the 3. But a nice boost vs main. Also on main when tossing the model around I see ~600% CPU with this it's ~100%.

So you may accept this PR?

**Comment by psobolewskiPhD** [comment]
> > With the latest commits everything works on all 3 versions of numpy I tested (1.26.1, 1.26.6, and 2.0.1), as fresh wheels from pypi. I dont see a difference in performance between the 3. But a nice boost vs main. Also on main when tossing the model around I see ~600% CPU with this it's ~100%.
> 
> So you may accept this PR?

I don't feel comfortable with a review, due to time and know-how. I've dismissed the issue as resolved.
The performance is definitely nice!

**Comment by Czaki** [comment]
@jni friendly reminder

**Review by jni — APPROVED** [review]
@Czaki I've updated the comments and code a bit. Hopefully I haven't broken anything. 😅 🙏 This is ready to go as far as I'm concerned. 🙏 Thank you for your persistence here, it's very cool and turned out to be annoyingly tricky!

**Comment by jni** [comment]
Grrr, it's ruff that removed the type: ignore comment.

File: napari/_check_numpy_version.py
Comment (Czaki) [inline napari/_check_numpy_version.py]:
```suggestion
    # Internally, it uses delocate, which stores the openblas
    # library in the .dylibs directory.
```

**Pull request merged**
